### PR TITLE
Xiao/LKG configuration cache

### PR DESCRIPTION
**Pull request by ciaozhang**
Introduce a LKG configuration cache to store each valid base configuration instead of a single entry of configuration.

### CODE REVIEW

**Comment by ciaozhang** [comment]
This PR is not ready for reivew.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+            else if (config1.Issuer == config2.Issuer && config1.SigningKeys.Count == config2.SigningKeys.Count
+                     && !config1.SigningKeys.Select(x => x.InternalId).Except(config1.SigningKeys.Select(x => x.InternalId)).Any())
+                return true;
+            else
+                return false;
+        }
+
+        public int GetHashCode(BaseConfiguration config)
+        {
+            return (config.Issuer + string.Join("", config.SigningKeys.Select(x => x.InternalId).OrderBy(x => x.FirstOrDefault()))).GetHashCode();
```
Comment (TimHannMSFT) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs]: this might work for the prototype but we shouldn't be doing string operations to produce the hash code in the release version. Can you at least comment what you plan to do here?

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    internal class BaseConfigurationComparer : IEqualityComparer<BaseConfiguration>
+    {
+        public bool Equals(BaseConfiguration config1, BaseConfiguration config2)
+        {
+            if (config1 == null && config2 == null)
+                return true;
+            else if (config1 == null || config2 == null)
+                return false;
+            else if (config1.Issuer == config2.Issuer && config1.SigningKeys.Count == config2.SigningKeys.Count
+                     && !config1.SigningKeys.Select(x => x.InternalId).Except(config1.SigningKeys.Select(x => x.InternalId)).Any())
```
Comment (TimHannMSFT) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs]: Is the InternalId some generated hash of the crypto artifact? How can we be sure this isn't duplicated with a valid and invalid artifact in config1 vs config2? #Resolved

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs [hunk 3]
```diff
@@ -76,9 +93,39 @@ public BaseConfiguration LastKnownGoodConfiguration
             {
                 _lastKnownGoodConfiguration = value ?? throw LogHelper.LogArgumentNullException(nameof(value));
                 _lastKnownGoodConfigFirstUse = DateTime.UtcNow;
+
+                if (_lkgConfigurationCache == null)
+                    _lkgConfigurationCache = new Dictionary<BaseConfiguration, DateTime>(BaseConfigurationComparer);
+                
+                _lkgConfigurationCache[_lastKnownGoodConfiguration] = DateTime.UtcNow + LastKnownGoodLifetime;
```
Comment (TimHannMSFT) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: >_lkgConfigurationCache [](http://example.com/codeflow?start=16&length=22)

Is there any sane upper bound we can add on this cache? Currently it's only bounded by the LastKnownGoodLifetime correct? This goes back to the complexity functionality tradeoff but seems reasonable to have a high upper bound here which can be overridden to prevent runaway cache in some error case where ESTS starts doing something weird with metadata publishing.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs [hunk 4]
```diff
@@ -76,9 +93,39 @@ public BaseConfiguration LastKnownGoodConfiguration
...
+
+                //remove expired configuration to avoid memory leak 
+                foreach (KeyValuePair<BaseConfiguration, DateTime> lkgConfiguration in _lkgConfigurationCache)
```
Comment (TimHannMSFT) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: >foreach [](http://example.com/codeflow?start=16&length=7)

Does this work? My memory of modifying a collection while iterating over it was that it will throw an exception.

[https://stackoverflow.com/questions/2347269/modifying-net-dictionary-while-enumerating-through-it](https://stackoverflow.com/questions/2347269/modifying-net-dictionary-while-enumerating-through-it)

Generally you'll want to create a collection and remove once finished with iteration.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs [hunk 5]
```diff
@@ -76,9 +93,39 @@ public BaseConfiguration LastKnownGoodConfiguration
...
+        /// <summary>
+        /// 
+        /// </summary>
+        /// <returns></returns>
+        public IDictionary<BaseConfiguration, DateTime> GetValidLkgConfiguraitonFromCache()
+        {
+            if (_lkgConfigurationCache == null)
+                return null;
+
+            foreach (KeyValuePair<BaseConfiguration, DateTime> lkgConfiguration in _lkgConfigurationCache)
```
Comment (TimHannMSFT) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: >foreach [](http://example.com/codeflow?start=12&length=7)

same comment regarding removing while iterating, consider a private helper since you're doing this in two places.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs (hunk 3 above)
Comment (ciaozhang) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: We could set the upper limit for the cache of lavage our EventBaseLruCache.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs (hunk 1 above)
Comment (ciaozhang) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs]: Here to use issuer and sorted signingKeys.InternalId to generate a hashcode.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs (hunk 2 above)
Comment (ciaozhang) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs]: Is the InternalId some generated hash of the crypto artifact?
Yes

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs (hunk 1 above)
Comment (TimHannMSFT) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs]: take a look at something like: [https://stackoverflow.com/questions/59375124/how-to-use-system-hashcode-combine-with-more-than-8-values](https://stackoverflow.com/questions/59375124/how-to-use-system-hashcode-combine-with-more-than-8-values)

We shouldn't be doing string operations in the GetHashCode method.

Comment (ciaozhang) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs]: Update the GetHashcode method with "^".

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 6]
```diff
@@ -1258,14 +1258,22 @@ private async Task<TokenValidationResult> ValidateTokenAsync(JsonWebToken jsonWe
                         }
                     }
 
-                    if (TokenUtilities.IsRecoverableConfiguration(validationParameters, currentConfiguration, out currentConfiguration))
+                    if (TokenUtilities.ShouldValidateWithLKG(validationParameters))
                     {
                         validationParameters.RefreshBeforeValidation = false;
                         validationParameters.ValidateWithLKG = true;
-                        tokenValidationResult = await ValidateTokenAsync(jsonWebToken, validationParameters, currentConfiguration).ConfigureAwait(false);
+                        var recoverableException = tokenValidationResult.Exception;
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:1263]: Do we need to create this local variable 'recoverableException', it is only used once.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs (hunk 2 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs]: The InternalId is based off this spec: https://www.rfc-editor.org/rfc/rfc7638

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs (hunk 6 above)
Comment (ciaozhang) [reply src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:1263]: It is make sure we are recovering from the current configuration. Otherwise, each time we loop through the cache might get different exception.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs [hunk 7]
```diff
@@ -63,6 +80,25 @@ public virtual Task<BaseConfiguration> GetBaseConfigurationAsync(CancellationTok
             throw new NotImplementedException();
         }
 
+        /// <summary>
+        /// Gets all valid last known good configurations from the cache.
+        /// </summary>
+        /// <returns>A collection of all valid last known good configurations.</returns>
+        internal ICollection<BaseConfiguration> GetValidLkgConfiguraitonFromCache()
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: Suggest dropping the postfix: 'Cache'

Comment (TimHannMSFT) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: Makes sense, a bit superfluous I suppose, LKG will always be a cache of some sort.

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs [hunk 8]
```diff
@@ -330,6 +331,43 @@ public void ResetLastKnownGoodLifetime()
             TestUtilities.AssertFailIfErrors(context);
         }
 
+        [Fact]
+        public void TestConfigurationComparer()
+        {
+            TestUtilities.WriteHeader($"{this}.TestConfigurationComparer", "TestConfigurationComparer", true);
+            var context = new CompareContext();
+
+            var config1 = new OpenIdConnectConfiguration() { TokenEndpoint = Default.Issuer + "oauth/token", Issuer = Default.Issuer };
```
Comment (TimHannMSFT) [inline test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs]: >TokenEndpoint = Default.Issuer + "oauth/token" [](http://example.com/codeflow?start=61&length=46)

here and below this will produce: "[http://Default.Issuer.comoauth/token](http://Default.Issuer.comoauth/token)"

add in the '/' for a more valid URI.

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs [hunk 9]
```diff
@@ -330,6 +331,43 @@ public void ResetLastKnownGoodLifetime()
...
+            var config2 = new OpenIdConnectConfiguration() { TokenEndpoint = Default.Issuer + "oauth/token", Issuer = Default.Issuer };
+            config2.SigningKeys.Add(KeyingMaterial.DefaultRsaSecurityKey1);
+            config2.SigningKeys.Add(KeyingMaterial.DefaultX509Key_2048);
+            config2.SigningKeys.Add(KeyingMaterial.DefaultRsaSecurityKey2);
+
+            var config3 = new OpenIdConnectConfiguration() { TokenEndpoint = Default.Issuer + "oauth/token", Issuer = Default.Issuer };
+            config3.SigningKeys.Add(KeyingMaterial.DefaultRsaSecurityKey1);
+
+            var config4 = new OpenIdConnectConfiguration() { TokenEndpoint = Default.Issuer + "oauth/token", Issuer = Default.Issuer + "1" };
+            config3.SigningKeys.Add(KeyingMaterial.DefaultRsaSecurityKey1);
```
Comment (TimHannMSFT) [inline test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs]: >config3 [](http://example.com/codeflow?start=12&length=7)

Is this supposed to be config4?

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs (hunk 8 above)
Comment (TimHannMSFT) [inline test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs]: >config1 [](http://example.com/codeflow?start=16&length=7)

Please add comments to be explicit about the test cases:

Config,
Config with same keys but different order
Config with overlapping key
Config with overlapping key but different issuer

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs [hunk 10]
```diff
@@ -330,6 +331,43 @@ public void ResetLastKnownGoodLifetime()
             TestUtilities.AssertFailIfErrors(context);
         }
 
+        [Fact]
+        public void TestConfigurationComparer()
```
Comment (TimHannMSFT) [inline test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs:371]: >TestConfigurationComparer [](http://example.com/codeflow?start=20&length=25)

Test case where token endpoint differs but is considered duplicate because it's not part of the hash check?

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs [hunk 11]
```diff
@@ -63,6 +80,25 @@ public virtual Task<BaseConfiguration> GetBaseConfigurationAsync(CancellationTok
...
+                return null;
+
+            var expiredLkgConfiguration = _lkgConfigurationCache.Where(x => x.Value < DateTime.UtcNow).ToArray();
+
+            foreach (KeyValuePair<BaseConfiguration, DateTime> lkgConfiguration in expiredLkgConfiguration)
+            {
+                _lkgConfigurationCache.Remove(lkgConfiguration.Key);
+            }
+
+            return _lkgConfigurationCache.Any() ? _lkgConfigurationCache.Keys : null;
```
Comment (TimHannMSFT) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: >null [](http://example.com/codeflow?start=80&length=4)

I think, as a developer, I'd expect this to return an empty set in this case not null. null seems to be the case where there's no cache (is that possible?) empty should be when there's nothing valid (e.g. all expired) IMO.

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs (hunk 9 above)
Comment (TimHannMSFT) [reply test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs]: Does anything real-world prevent duplicate keys being added? Here it looks like this Config results in SigningKeys collection with duplicate keys, that's something we'd never want in practice, correct?

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs [hunk 12]
```diff
@@ -330,6 +331,43 @@ public void ResetLastKnownGoodLifetime()
...
+            var configurationManager = new MockConfigurationManager<OpenIdConnectConfiguration>(config1, config1);
+            IdentityComparer.AreEqual(configurationManager.GetValidLkgConfiguraitonFromCache().Count, 1, context);
+
+            configurationManager.LastKnownGoodConfiguration = config2;
+            IdentityComparer.AreEqual(configurationManager.GetValidLkgConfiguraitonFromCache().Count, 1, context);
+
+            configurationManager.LastKnownGoodConfiguration = config3;
+            IdentityComparer.AreEqual(configurationManager.GetValidLkgConfiguraitonFromCache().Count, 2, context);
+
+            configurationManager.LastKnownGoodConfiguration = config4;
```
Comment (TimHannMSFT) [inline test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs]: >LastKnownGoodConfiguration [](http://example.com/codeflow?start=33&length=26)

because of the testing bug, we're adding a config with no keys to the cache, would we ever want to actually do this? Seems like something should prevent this, I don't see what the value would be. Perhaps there should be a warning log too.

File: test/Microsoft.IdentityModel.TestUtils/MockConfigurationManager.cs [hunk 13]
```diff
@@ -74,6 +74,30 @@ public MockConfigurationManager(T configuration, T lkgConfiguration, T refreshed
...
+            LastKnownGoodLifetime = lkgLifetime;
+        }
+
+        /// <summary>
+        /// Initializes an new instance of <see cref="MockConfigurationManager{T}"/> with a Configuration instance and a LKG Configuration instance.
+        /// </summary>
+        /// <param name="configuration">Configuration of type OpenIdConnectConfiguration or WsFederationConfiguration.</param>
+        /// <param name="lkgConfiguration">Configuration of type OpenIdConnectConfiguration or WsFederationConfiguration.</param>
+        /// <param name="lkgLifetime">The LKG configuration lifetime.</param>
+        public MockConfigurationManager(T configuration, T lkgConfiguration, TimeSpan lkgLifetime) : this(configuration, lkgLifetime)
```
Comment (TimHannMSFT) [inline test/Microsoft.IdentityModel.TestUtils/MockConfigurationManager.cs:99]: >T [](http://example.com/codeflow?start=57&length=1)

BaseConfiguration?

**Comment by TimHannMSFT** [comment]
>     /// <typeparam name="T">must be a class.</typeparam>

Not part of your change but why do we need to be so lax here? Wouldn't we always want T to be of type BaseConfiguration?

---
Refers to: test/Microsoft.IdentityModel.TestUtils/MockConfigurationManager.cs:17 in 32e63b8. [](commit_id = 32e63b88faf74433cc2ca7f8899836b64ad687c6, deletion_comment = False)

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs (hunk 1 above)
Comment (TimHannMSFT) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs]: Are you doing force pushes here? (I see this changed but it still shows as the first iteration) Once you push a PR, please just push new changes as new commits in case anyone has checked out your branch and is attached to its change history.

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs (hunk 10 above)
Comment (TimHannMSFT) [reply test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs:371]: Also consider a test case which covers duplicate keys with same kid

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs [hunk 14]
```diff
@@ -330,6 +331,43 @@ public void ResetLastKnownGoodLifetime()
             TestUtilities.AssertFailIfErrors(context);
         }
```
Comment (TimHannMSFT) [inline test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs:368]: >} [](http://example.com/codeflow?start=8&length=1)

Unit tests for aging out the LKG entries?

**Review by TimHannMSFT — CHANGES_REQUESTED** [review]
:clock1:

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs (hunk 11 above)
Comment (ciaozhang) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: updated

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs (hunk 10 above)
Comment (ciaozhang) [reply test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs:371]: Added test

File: test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs (hunk 12 above)
Comment (ciaozhang) [reply test/Microsoft.IdentityModel.Protocols.OpenIdConnect.Tests/ConfigurationManagerTests.cs]: We have ConfigurationValidtor to make sure there is at least 1 key in the key set.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs [hunk 15]
```diff
@@ -76,6 +111,24 @@ public BaseConfiguration LastKnownGoodConfiguration
             {
                 _lastKnownGoodConfiguration = value ?? throw LogHelper.LogArgumentNullException(nameof(value));
                 _lastKnownGoodConfigFirstUse = DateTime.UtcNow;
+
+                if (_lkgConfigurationCache == null)
+                    _lkgConfigurationCache = new EventBasedLRUCache<BaseConfiguration, DateTime>(LastKnownGoodConfigurationSizeLimit, TaskCreationOptions.None, BaseConfigurationComparer, true, maintainLRU: true);
```
Comment (ciaozhang) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: Adopt the EventBaseLRUCache here. @brentschmaltz @dannybtsai 
What is a reasonable value for default cache size limit?

File: test/Microsoft.IdentityModel.TestUtils/MockConfigurationManager.cs (hunk 13 above)
Comment (ciaozhang) [reply test/Microsoft.IdentityModel.TestUtils/MockConfigurationManager.cs:99]: the type should be derived from BaseConfiguration

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
```
Comment (RojaEnnam) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs]: This should be config2?
config1.SigningKeys.Select(x => x.InternalId).Except(**config2**.SigningKeys.Select(x => x.InternalId)).Any())

Comment (ciaozhang) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationComparer.cs]: Good catch!

File: src/Microsoft.IdentityModel.Tokens/TokenUtilities.cs [hunk 17]
```diff
@@ -190,7 +190,6 @@ internal static bool IsRecoverableException(Exception exception)
             // using 'GetType()' instead of 'is' as SecurityTokenUnableToValidException (and others) extend SecurityTokenInvalidSignatureException
             // we want to make sure that the clause for SecurityTokenUnableToValidateException is hit so that the ValidationFailure is checked
             return exception.GetType().Equals(typeof(SecurityTokenInvalidSignatureException))
-                   || exception is SecurityTokenInvalidSigningKeyException
```
Comment (ciaozhang) [inline src/Microsoft.IdentityModel.Tokens/TokenUtilities.cs:193]: TBD: SecurityTokenInvalidSigningKeyException is recoverable exception or not.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs (hunk 15 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: 10 - 20

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs [hunk 18]
```diff
@@ -63,6 +86,18 @@ public virtual Task<BaseConfiguration> GetBaseConfigurationAsync(CancellationTok
             throw new NotImplementedException();
         }
 
+        /// <summary>
+        /// Gets all valid last known good configurations.
+        /// </summary>
+        /// <returns>A collection of all valid last known good configurations.</returns>
+        internal ICollection<BaseConfiguration> GetValidLkgConfiguraitons()
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: Please rename: GetValidLkgConfiguraitons => GetValidLkgConfigurations.

'i' is in the wrong place.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs [hunk 19]
```diff
@@ -20,6 +22,10 @@ public abstract class BaseConfigurationManager
         private BaseConfiguration _lastKnownGoodConfiguration;
         private DateTime? _lastKnownGoodConfigFirstUse = null;
 
+        private EventBasedLRUCache<BaseConfiguration, DateTime> _lkgConfigurationCache = null;
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: be consistent with naming.
_lkgConfigurationCache vs. _lastKnownGoodConfigFirstUse vs. _lastKnownGoodConfiguration vs. _lastKnownGoodConfigurationSizeLimit

_lkgConfiguration is my choice.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs [hunk 20]
```diff
@@ -76,6 +111,24 @@ public BaseConfiguration LastKnownGoodConfiguration
             {
                 _lastKnownGoodConfiguration = value ?? throw LogHelper.LogArgumentNullException(nameof(value));
                 _lastKnownGoodConfigFirstUse = DateTime.UtcNow;
+
+                if (_lkgConfigurationCache == null)
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: I don't see a reason to delay the creation of the cache, might as well new it up in the ctor or assign the private variable.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs (hunk 15 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: I don't think we should set maintainLRU to true.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs [hunk 21]
```diff
@@ -63,6 +86,18 @@ public virtual Task<BaseConfiguration> GetBaseConfigurationAsync(CancellationTok
...
+        /// <summary>
+        /// Gets all valid last known good configurations.
+        /// </summary>
+        /// <returns>A collection of all valid last known good configurations.</returns>
+        internal ICollection<BaseConfiguration> GetValidLkgConfiguraitons()
+        {
+            if (_lkgConfigurationCache == null)
+                return null;
+
+            return _lkgConfigurationCache.ToArray().Where(x => x.Value.Value > DateTime.UtcNow).Select(x => x.Key).ToArray();
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: The EventBasedLRUCache will keep the cache the right size and remove expired items in the background.
Could we just return _LkgConfigurationCache.ToArray() assuming all items in the cache should be tried?

If we make the other change to instantiate the cache in the ctor, no need to check for null.

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 22]
```diff
@@ -1258,14 +1258,22 @@ private async Task<TokenValidationResult> ValidateTokenAsync(JsonWebToken jsonWe
...
                         validationParameters.RefreshBeforeValidation = false;
                         validationParameters.ValidateWithLKG = true;
-                        tokenValidationResult = ValidateToken(jsonWebToken, validationParameters, currentConfiguration);
+                        var recoverableException = tokenValidationResult.Exception;
 
-                        if (tokenValidationResult.IsValid)
-                            return tokenValidationResult;
+                        foreach (BaseConfiguration lkgConfiguration in validationParameters.ConfigurationManager.GetValidLkgConfiguraitons())
+                        {
+                            if (!lkgConfiguration.Equals(currentConfiguration) && TokenUtilities.IsRecoverableConfiguration(jsonWebToken.Kid, currentConfiguration, lkgConfiguration, recoverableException))
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs:1267]: We checked above if the error is recoverable.
Could we just try all the configurations in the cache?

We only add configurations to the cache if they are different.
We could have a situation where we have configurations that are different only by issuer with the same keys, but i think it is ok as we are in a temporary situation and keeping the logic simple is a good choice.

File: src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs [hunk 23]
```diff
@@ -1258,14 +1258,22 @@ private async Task<TokenValidationResult> ValidateTokenAsync(JsonWebToken jsonWe
                         }
                     }
 
-                    if (TokenUtilities.IsRecoverableConfiguration(validationParameters, currentConfiguration, out currentConfiguration))
+                    if (TokenUtilities.ShouldValidateWithLKG(validationParameters))
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.JsonWebTokens/JsonWebTokenHandler.cs]: If we make the change in BaseConfigurationManager.GetValidLkgConfiguraitons(), TokenUtilities.ShouldValidateWithLKG is not needed as we assume all entries in the lkgCache should be tried.
If there are none, an empty array will be returned.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs (hunk 19 above)
Comment (ciaozhang) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: Going to use _lastKnownGoodConfigurationCache and it is consistent with other public property.

File: src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs (hunk 15 above)
Comment (ciaozhang) [reply src/Microsoft.IdentityModel.Tokens/BaseConfigurationManager.cs]: Use defalut value(false) for maintainLRU

File: src/Microsoft.IdentityModel.Tokens/TokenUtilities.cs (hunk 17 above)
Comment (brentschmaltz) [reply src/Microsoft.IdentityModel.Tokens/TokenUtilities.cs:193]: I don't think SecurityTokenInvalidSigningKeyException is recoverable.

File: src/Microsoft.IdentityModel.Tokens/TokenUtilities.cs [hunk 24]
```diff
@@ -200,30 +199,39 @@ internal static bool IsRecoverableException(Exception exception)
...
+        /// <param name="lkgConfiguration">The LKG exception to check.</param>
+        /// <param name="currentException">The exception to check.</param>
         /// <returns><c>true</c> if the configuration is recoverable otherwise, <c>false</c>.</returns>
-        internal static bool IsRecoverableConfiguration(TokenValidationParameters validationParameters, BaseConfiguration configuration, out BaseConfiguration currentConfiguration)
+        internal static bool IsRecoverableConfiguration(string kid, BaseConfiguration currentConfiguration, BaseConfiguration lkgConfiguration, Exception currentException)
         {
-            bool isRecoverableConfiguration = (validationParameters.ConfigurationManager.UseLastKnownGoodConfiguration
-                && validationParameters.ConfigurationManager.LastKnownGoodConfiguration != null
-                && !ReferenceEquals(configuration, validationParameters.ConfigurationManager.LastKnownGoodConfiguration));
+            Lazy<bool> isRecoverableIssuer = new Lazy<bool>(() => currentConfiguration.Issuer != lkgConfiguration.Issuer);
```
Comment (brentschmaltz) [inline src/Microsoft.IdentityModel.Tokens/TokenUtilities.cs]: I am not understanding the need for this Lazy.
On line 214, why not just write: return (currentConfiguration.Issuer != lkgConfiguration.Issuer)?

While the code may seem cleaner, Lazy introduces a new object creation.

**Review by brentschmaltz — APPROVED** [review]

File: src/Microsoft.IdentityModel.Tokens/TokenUtilities.cs (hunk 24 above)
Comment (ciaozhang) [reply src/Microsoft.IdentityModel.Tokens/TokenUtilities.cs]: resolved
